File: backends/java/zappy-domain/src/main/java/com/zappymart/domain/accounts/Wishlist.java
package com.zappymart.domain.accounts;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public record Wishlist(String ownerId, List<String> productIds) {

    public Wishlist {
        Objects.requireNonNull(ownerId, "ownerId");
        productIds = List.copyOf(productIds);
    }

    public static Wishlist emptyFor(String ownerId) {
        return new Wishlist(ownerId, List.of());
    }

    public Wishlist wishing(String productId) {
        if (productIds.contains(productId)) {
            return this;
        }
        List<String> wished = new ArrayList<>();
        wished.add(productId);
        wished.addAll(productIds);
        return new Wishlist(ownerId, wished);
    }

    public Wishlist noLongerWishing(String productId) {
        if (!productIds.contains(productId)) {
            return this;
        }
        List<String> wished = new ArrayList<>(productIds);
        wished.remove(productId);
        return new Wishlist(ownerId, wished);
    }

    public Wishlist absorbing(Wishlist other) {
        Wishlist merged = this;
        for (String productId : other.productIds().reversed()) {
            merged = merged.wishing(productId);
        }
        return merged;
    }
}

File: backends/kotlin/zappy-domain/src/main/kotlin/nl/zappymart/domain/accounts/Wishlist.kt
package nl.zappymart.domain.accounts

data class Wishlist(val ownerId: String, val productIds: List<String>) {

    fun with(productId: String): Wishlist =
        if (productIds.contains(productId)) this else copy(productIds = listOf(productId) + productIds)

    fun without(productId: String): Wishlist =
        copy(productIds = productIds.filterNot { saved -> saved == productId })

    fun mergedWith(other: Wishlist): Wishlist =
        other.productIds.reversed().fold(this) { running, productId -> running.with(productId) }

    companion object {
        fun emptyFor(ownerId: String) = Wishlist(ownerId, emptyList())
    }
}
